articles/intraprocess_communication.md

```diff
@@ -0,0 +1,437 @@
+4. Then the `IntraProcessManager` is notified about the existence of the new `Publisher` through the method `IntraProcessManager::add_publisher(PublisherBase::SharedPtr publisher, PublisherOptions options)`.
+5. `IntraProcessManager::add_publisher(...)` stores the `Publisher` information in an internal structure of type `PublisherInfo`. The structure contains information about the `Publisher`, such as its QoS and its topic name, and a weak pointer for the
+   `Publisher` object. An `uint64_t pub_id` unique within the `rclcpp::Context` is assigned to the `Publisher`. The `IntraProcessManager` contains a `std::map<uint64_t, PublisherInfo>` object where it is possible to retrieve the `PublisherInfo` of a
+   specific `Publisher` given its id. The function returns the `pub_id`, that is stored within the `Publisher`.
+
+If the `Publisher` QoS is set to `transient local`, then the `Publisher::SetupIntraProcess(...)` method will also create a ring buffer of the size specified by the depth from the QoS.
+
+### Creating a subscription
+
+1. User calls `Node::create_subscription<MessageT>(...)`.
+2. This boils down to `NodeTopics::create_subscription(...)`, where a `Subscription` is created through the factory.
+3. Here, if intra-process communication is enabled, intra-process related variables are initialized through the `Subscription::SetupIntraProcess(...)` method. The most relevant ones being the ring buffer and the waitable object.
+4. Then the `IntraProcessManager` is notified about the existence of the new `Subscription` through the method `IntraProcessManager::add_subscription(SubscriptionBase::SharedPtr subscription, SubscriptionOptions options)`.
+5. `IntraProcessManager::add_subscription(...)` stores the `Subscription` information in an internal structure of type `SubscriptionInfo`. The structure contains information about the `Subscription`, such as its QoS, its topic name and the type of
+   its callback, and a weak pointer for the `Subscription` object. An `uint64_t sub_id` unique within the `rclcpp::Context` is assigned to the `Subscription`. The `IntraProcessManager` contains a `std::map<uint64_t, SubscriptionInfo>` object where it
+   is possible to retrieve the `SubscriptionInfo` of a specific `Subscription` given its id. There is also an additional structure `std::map<uint64_t, std::pair<std::set<uint64_t>, std::set<uint64_t>>>`. The key of the map is the unique id of a
+   `Publisher`. The value of the map is a pair of sets of ids. These sets contain the ids of the `Subscription`s that can communicate with the `Publisher`. We have two different sets because we want to differentiate the `Subscription`s depending on
+   whether they request ownership of the received messages or not.
+6. The `SubscriptionIntraProcessWaitable` object is added to the list of Waitable interfaces of the node through `node_interfaces::NodeWaitablesInterface::add_waitable(...)`.
+
+### Publishing only intra-process
+
+#### Publishing unique_ptr
```

Comment:
Sometimes there's something to be said for [using UML diagrams](http://plantuml.com/).

Reply:
Sorry, but I don't get what do you mean

Reply:
@alsora I am guessing the comment is recommending UML based diagrams to describe visually some of the sections of this PR. For. Eg, Creating a publisher,  Creating a subscription,  Publishing only intra-process etc can have [sequence diagrams](http://plantuml.com/sequence-diagram) in addition to the description.